crates/isograph_compiler/src/watch.rs

```diff
@@ -52,20 +64,106 @@ pub async fn handle_watch_command(
     .await
 }
 
-fn any_modified_path_is_outside_artifact_directory(
+fn changed_files(
     events: &[DebouncedEvent],
-    artifact_directory: &PathBuf,
-) -> bool {
-    for event in events.iter() {
-        for watched_path in event.paths.iter() {
-            // These paths are canonicalized, so it is sufficient
-            // to check starts_with
-            if !watched_path.starts_with(artifact_directory) {
-                return true;
+    config: &CompilerConfig,
+) -> Option<Vec<SourceFileEvent>> {
+    let mut source_file_events = vec![];
+    for event in events {
+        if let Some(source_file_event) = match event.kind {
+            EventKind::Create(create_kind) => {
+                process_create_event(config, create_kind, &event.paths)
+            }
+            EventKind::Modify(modify_kind) => {
+                process_modify_event(config, modify_kind, &event.paths)
             }
+            EventKind::Remove(remove_kind) => {
+                process_remove_event(config, remove_kind, &event.paths)
+            }
+            _ => None,
+        } {
+            source_file_events.push(source_file_event);
         }
     }
-    false
+    if source_file_events.is_empty() {
+        None
+    } else {
+        Some(source_file_events)
+    }
+}
+
+fn process_create_event(
+    config: &CompilerConfig,
+    create_kind: CreateKind,
+    paths: &[PathBuf],
+) -> Option<SourceFileEvent> {
+    match create_kind {
+        CreateKind::File if paths.len() == 1 => changed_file_kind(config, &paths[0])
+            .map(|file_kind| (SourceEventKind::CreateOrModify(paths[0].clone()), file_kind)),
+        _ => None,
+    }
+}
+
+fn process_modify_event(
+    config: &CompilerConfig,
+    modify_kind: ModifyKind,
+    paths: &[PathBuf],
+) -> Option<SourceFileEvent> {
+    match modify_kind {
+        ModifyKind::Data(_) if paths.len() == 1 => changed_file_kind(config, &paths[0])
+            .map(|file_kind| (SourceEventKind::CreateOrModify(paths[0].clone()), file_kind)),
+        ModifyKind::Name(rename_mode) => {
+            match rename_mode {
+                // This event could be fired once on delete or twice on rename
+                RenameMode::Any if paths.len() == 1 => {
+                    changed_file_kind(config, &paths[0]).map(|file_kind| {
+                        if paths[0].exists() {
+                            (SourceEventKind::CreateOrModify(paths[0].clone()), file_kind)
+                        } else {
+                            (SourceEventKind::Remove(paths[0].clone()), file_kind)
+                        }
+                    })
+                }
+                RenameMode::Both if paths.len() == 2 => {
```

Comment:
maybe unrelated, but if we move an entire folder compilation breaks (once)

```
mkdir folder
touch folder/file
mv ./folder ./demos/pet-demo/src/components
```

```
Unable to read the file at the following path: "/Users/rbalicki/code/isograph/demos/pet-demo/src/components/folder".
Reason: Is a directory (os error 21)
```

fixes itself after modifying a file though